FlowCal/gate.py

```diff
@@ -334,11 +352,19 @@ def density2d(data, channels=[0,1],
     if data_ch.shape[0] <= 1:
         raise ValueError('data should have more than one event')
 
-    # Extract default bins if necessary
-    if (bins is None and hasattr(data_ch, 'hist_bin_edges')
-            and data_ch.hist_bin_edges(0) is not None
-            and data_ch.hist_bin_edges(1) is not None):
-        bins = np.array(data_ch.hist_bin_edges())
+    # If bins are not specified, try to get bins from data object
+    if (hasattr(data_ch, 'hist_bins')
+            and data_ch.hist_bins(0) is not None
+            and data_ch.hist_bins(1) is not None):
+        # Duplicate bins into a two-element list if necessary
+        if not isinstance(bins, list):
+            bins = [bins, bins]
+        # X Axis
+        if bins[0] is None or isinstance(bins[0], int):
+            bins[0] = data_ch.hist_bins(0, nbins=bins[0], log=xlog)
+        # Y axis
+        if bins[1] is None or isinstance(bins[1], int):
+            bins[1] = data_ch.hist_bins(1, nbins=bins[1], log=ylog)
```

Comment:
Will add `channels=0` to the `hist_bins` call. Read above for responses to your other comments.